File: utils/lassoClasifier/lassoLinearSVM.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Copyright (C) 2016 Gamaleldin F. Elsayed and Farzaneh Najafi
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Gamaleldin F. Elsayed
% 
% lassoLinearSVM.m
%
% This function performs linear SVM classification with l1 penalty. The 
% l1 regularization parameter is obtain by kfold cross validation with 1SE
% criteria. This function interfaces and calls a python function 
% (lassoLinearSVM.py), which does all the heavy-duty work.
%
% Inputs:
%   X: features matrix of size (trials x neurons).
%   Y: labels vector of size (trials x 1).
%   kfold: cross-validation folds.
% Outputs:
%   SVMl1Model: structure with the following fields
%   	.Model: the scalar slope of the regression model
%           .Beta: linear svm coefficients of size (neuorns x 1).
%           .Bias: bias component.
%           .regularization: l1-regularization parameter used.
%           .CVerror: Cross validation error of the model
%           .Object: svm object passed from the python function
%   	.Standard: values used in feature normalization and scaling
%           .meanX: center features to have zero mean.
%           .stdX: scale features to have unity std.
%   	.CrossValidation: field that contains information the cross
%   	validation proceudure used to find the regularization parameters.
%           .kfold: kfold used in cross validation.
%           .regularization: regularization valused examined.
%           .meanCVerror: mean cross validation error from 100 runs. Different entries reflect result from different values of regularization specified above.
%           .semCVerror: standard error of the mean of cross validation
%           error from 100 runs. Different entries reflect result from different values of regularization specified above.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [SVMl1Model] = lassoLinearSVM(X, Y, kfold)

% add path to python. You need to be in the directory of lassoLinearSVM
% functions.
if count(py.sys.path,'') == 0
insert(py.sys.path,int32(0),'');
end

outputs = py.lassoLinearSVM.lassoLinearSVM(X(:).', Y(:).', kfold);
w = cell2mat(cell(py.list(outputs{1})));
b = cell2mat(cell(py.list(outputs{2})));
bestCVerror = outputs{3};
bestc = outputs{4};
linear_svm = outputs{5};
meanCVerror = cell2mat(cell(py.list(outputs{6})));
semCVerror = cell2mat(cell(py.list(outputs{7})));
c = cell2mat(cell(py.list(outputs{8})));
meanX = cell2mat(cell(py.list(outputs{9})));
stdX = cell2mat(cell(py.list(outputs{10})));

figure
subplot(211)
hold on
errorbar(c, meanCVerror,  semCVerror, 'k')
plot(bestc, bestCVerror, 'ro')
hold off
ylim([0 1])
set(gca, 'xscale', 'log')
xlabel('inverse of lasso-regularization parameter (c)')    
ylabel('cross-validation error')    
subplot(212)
plot(sort(abs(w), 'descend'))
xlabel('ordered neurons')
ylabel('|w|')

SVMl1Model.Model.Beta = w;
SVMl1Model.Model.Bias = b;
SVMl1Model.Model.regularization = bestc;
SVMl1Model.Model.CVerror = bestCVerror;
SVMl1Model.Model.Object = linear_svm;
SVMl1Model.Standard.meanX = meanX;
SVMl1Model.Standard.stdX = stdX;
SVMl1Model.CrossValidation.kfold = kfold;
SVMl1Model.CrossValidation.regularization = c;
SVMl1Model.CrossValidation.meanCVerror = meanCVerror;
SVMl1Model.CrossValidation.semCVerror = semCVerror;

end

File: utils/lassoClasifier/lassoLinearSVM.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 15 10:41:20 2016

@author: gamalamin
"""

# -*- coding: utf-8 -*-
"""
Created on Wed May 18 12:54:46 2016

@author: gamalamin
"""
# X and Y are two vectors
def lassoLinearSVM(X, Y, kfold):
    #%% initialization ans imports
    import numpy as np
    import numpy.random as rng
    from sklearn import svm
    import matplotlib.pyplot as plt
    from scipy import stats
    
    
    Y = np.squeeze(np.array(Y).astype(int));
    L = len(Y);
    N = len(X)/L;
    kfold = int(kfold);    
    
    X = np.reshape(np.array(X.astype('float')), (L, N), order='F');

    #%% feature normalization and scale
    meanX = np.mean(X, axis = 0);
    stdX = np.std(X, axis = 0);
    X = (X-meanX)/stdX;
    
    #%% define functions
    def optimizeSVMReg(X, Y, kfold):
        numSamples = 100;
        c = 10**np.array(np.arange(-5, 3, 0.1))
        CVerror = np.ones((len(c), numSamples))
        for i in range(len(c)):
            for j in range(numSamples):
                CVerror[i, j] = CVSVM(X, Y, c[i], kfold);
        meanCVerror = np.mean(CVerror, axis = 1);
        semCVerror = stats.sem(CVerror, axis = 1);    
        ix = np.argmin(meanCVerror)
       # best error using standard error criteria
        msk = meanCVerror<=meanCVerror[ix]+semCVerror[ix];    
        bestc = c[msk][0]
        bestCVerror = meanCVerror[msk][0]
        plt.figure()
        plt.errorbar(c, meanCVerror,  yerr=semCVerror, color = 'k')
        plt.plot(bestc, bestCVerror, 'ro')
        plt.xscale('log')
        plt.xlabel('inverse lasso regularization parameter (c)')    
        plt.ylabel('cross-validation error')    
        return bestc, bestCVerror, meanCVerror, semCVerror, c
       
    def CVSVM(X, Y, c, kfold):
        L = len(np.squeeze(Y));
        ix = rng.permutation(np.arange(L))
        X = X[ix, :];
        Y = Y[ix];
        XCV = X[range(0, L/kfold), :]
        YCV = Y[range(0, L/kfold)]
        
        XTrain = X[range(L/kfold, L), :]
        YTrain = Y[range(L/kfold, L)]
        
        # Create a classifier: a support vector classifier
        linear_svm = svm.LinearSVC(C = c, loss='squared_hinge', penalty='l1', dual=False)
        linear_svm.fit(XTrain, np.squeeze(YTrain))
        CVerror = predictError(YCV, linear_svm.predict(XCV));
        return CVerror
        
    def predictError(Y, Yhat):
        error = sum(abs(np.squeeze(Yhat).astype(float)-np.squeeze(Y).astype(float)))/len(Y)
        return error
        
    #%% identify the best reg parameter using 1SE criteria
    bestc, bestCVerror, meanCVerror, semCVerror, c = optimizeSVMReg(X, Y, kfold)
    
    #%% run the linear SVM classifer with best regularization parameter
    # Create a classifier: a support vector classifier
    linear_svm = svm.LinearSVC(C = bestc, loss='squared_hinge', penalty='l1', dual=False)    
    linear_svm.fit(X, Y)
    wl1 =  np.squeeze(linear_svm.coef_);
    b = linear_svm.intercept_;
    
    return wl1, b, bestCVerror, bestc, linear_svm, meanCVerror, semCVerror, c, meanX, stdX
